tests/test_model.py

```diff
@@ -18,435 +18,439 @@
-
-        models = [exact_model, constraint, num_model, ode_model,
-                  connected_num_model]
-        for model in models:
-            new_model = pickle.loads(pickle.dumps(model))
-            # Compare signatures
-            self.assertEqual(model.__signature__, new_model.__signature__)
-            # Trigger the cached vars because we compare `__dict__` s
-            model.vars
-            new_model.vars
-            # Explicitly make sure the connectivity mapping is identical.
-            self.assertEqual(model.connectivity_mapping,
-                             new_model.connectivity_mapping)
-            if not isinstance(model, ODEModel):
-                model.function_dict
-                model.vars_as_functions
-                new_model.function_dict
-                new_model.vars_as_functions
-            self.assertEqual(model.__dict__, new_model.__dict__)
-
-    def test_MatrixSymbolModel(self):
-        """
-        Test a model which is defined by ModelSymbols, see #194
-        """
-        N = Symbol('N', integer=True)
-        M = MatrixSymbol('M', N, N)
-        W = MatrixSymbol('W', N, N)
-        I = MatrixSymbol('I', N, N)
-        y = MatrixSymbol('y', N, 1)
-        c = MatrixSymbol('c', N, 1)
-        a, b = parameters('a, b')
-        z, x = variables('z, x')
+def test_CallableNumericalModel2D():
+    """
+    Apply a CallableNumericalModel to 2D data, to see if it is
+    agnostic to data shape.
+    """
+    shape = (30, 40)
+
+    def function(a, b):
+        out = np.ones(shape) * a
+        out[15:, :] += b
+        return out
+
+    a, b = parameters('a, b')
+    y, = variables('y')
+
+    model = CallableNumericalModel({y: function}, [], [a, b])
+    data = 15 * np.ones(shape)
+    data[15:, :] += 20
+
+    fit = Fit(model, y=data)
+    fit_result = fit.execute()
+    assert fit_result.value(a) == pytest.approx(15)
+    assert fit_result.value(b) == pytest.approx(20)
+
+    def flattened_function(a, b):
+        out = np.ones(shape) * a
+        out[15:, :] += b
+        return out.flatten()
+
+    model = CallableNumericalModel({y: flattened_function}, [], [a, b])
+    data = 15 * np.ones(shape)
+    data[15:, :] += 20
+    data = data.flatten()
+
+    fit = Fit(model, y=data)
+    flat_result = fit.execute()
+
+    assert fit_result.value(a) == pytest.approx(flat_result.value(a))
+    assert fit_result.value(b) == pytest.approx(flat_result.value(b))
+
+    if fit_result.stdev(a) is None and flat_result.stdev(a) is None:
+        assert True
+    elif fit_result.stdev(a) is None or flat_result.stdev(a) is None:
+        assert False
+    else:
+        assert fit_result.stdev(a) == pytest.approx(flat_result.stdev(a))
+    
+    if fit_result.stdev(b) is None and flat_result.stdev(b) is None:
+        assert True
+    elif fit_result.stdev(b) is None or flat_result.stdev(b) is None:
+        assert False
+    else:
+        assert fit_result.stdev(b) == pytest.approx(flat_result.stdev(b))
```

Comment:
The same thing applies here: why is this suddenly split out between None or not None? The result should always be deterministic, if not we have other problems. So only the pytest.approx should be enough here.

Reply:
I tried to compare it with `unittest`. If both stdevs are None, there are equal. For `unittest` it's possible to compare it, but pytest doesn't (like you said abowe). In this spesific test there is a case, where both result are `None`. It was in my eyes a good way to compensate for the missing feature of pytest.

Reply:
Good point. But then I think that in hindsight a better test would be to just check that it is really `None`, since for `CallableNumericalModel` this is what we would expect. If we somehow improve that in the future than we will just just change it then.

So could you change this to an explicit check that they are all `None`?

Reply:
So you suggest checking if the `CallableNumericalModel` is `None` right? And if so, to skip the test with this model?

Reply:
No, I mean changing the relevant test to:
```python
assert fit_result.stdev(b) is None and flat_result.stdev(b) is None
```
to actually not allow anything else.